// @redis/client/dist/lib/client/parser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BasicCommandParser = exports.prefixKeys = exports.prefixKey = void 0;
/**
 * Prepends `keyPrefix` to `key`.
 *
 * Returns `key` unchanged when `keyPrefix` is `undefined` (the no-prefix hot path).
 * Uses a plain string concatenation when both sides are strings, and falls back to
 * `Buffer.concat` when either side is a `Buffer`.
 */
function prefixKey(keyPrefix, key) {
    if (keyPrefix === undefined)
        return key;
    if (typeof keyPrefix === 'string' && typeof key === 'string') {
        return keyPrefix + key;
    }
    return Buffer.concat([
        typeof keyPrefix === 'string' ? Buffer.from(keyPrefix) : keyPrefix,
        typeof key === 'string' ? Buffer.from(key) : key
    ]);
}
exports.prefixKey = prefixKey;
/**
 * Applies {@link prefixKey} to every key in a variadic argument, returning a new array.
 *
 * Used by the few commands that build their wire arguments outside of `parseCommand`
 * (e.g. `WATCH`), so they still honor the configured `keyPrefix`.
 */
function prefixKeys(keyPrefix, keys) {
    return Array.isArray(keys)
        ? keys.map(key => prefixKey(keyPrefix, key))
        : [prefixKey(keyPrefix, keys)];
}
exports.prefixKeys = prefixKeys;
class BasicCommandParser {
    #redisArgs = [];
    #keys = [];
    #keyPrefix;
    preserve;
    /**
     * @param keyPrefix Optional prefix prepended to every key pushed via
     * `pushKey`/`pushKeys`/`pushKeysLength`. Empty values are treated as "no prefix".
     */
    constructor(keyPrefix) {
        this.#keyPrefix = keyPrefix === undefined || keyPrefix.length === 0 ? undefined : keyPrefix;
    }
    get redisArgs() {
        return this.#redisArgs;
    }
    get keys() {
        return this.#keys;
    }
    get firstKey() {
        return this.#keys[0];
    }
    get cacheKey() {
        const tmp = new Array(this.#redisArgs.length * 2);
        for (let i = 0; i < this.#redisArgs.length; i++) {
            tmp[i] = this.#redisArgs[i].length;
            tmp[i + this.#redisArgs.length] = this.#redisArgs[i];
        }
        return tmp.join('_');
    }
    push(...arg) {
        this.#redisArgs.push(...arg);
    }
    ;
    pushVariadic(vals) {
        if (Array.isArray(vals)) {
            for (const val of vals) {
                this.push(val);
            }
        }
        else {
            this.push(vals);
        }
    }
    pushVariadicWithLength(vals) {
        if (Array.isArray(vals)) {
            this.#redisArgs.push(vals.length.toString());
        }
        else {
            this.#redisArgs.push('1');
        }
        this.pushVariadic(vals);
    }
    pushVariadicNumber(vals) {
        if (Array.isArray(vals)) {
            for (const val of vals) {
                this.push(val.toString());
            }
        }
        else {
            this.push(vals.toString());
        }
    }
    /**
     * Registers a key as both a routing key (`keys`) and a wire argument (`redisArgs`),
     * prepending `keyPrefix` when `applyPrefix` is true.
     */
    #addKey(key, applyPrefix) {
        const finalKey = applyPrefix ? prefixKey(this.#keyPrefix, key) : key;
        this.#keys.push(finalKey);
        this.#redisArgs.push(finalKey);
    }
    pushKey(key, applyPrefix = true) {
        this.#addKey(key, applyPrefix);
    }
    pushKeysLength(keys) {
        if (Array.isArray(keys)) {
            this.#redisArgs.push(keys.length.toString());
        }
        else {
            this.#redisArgs.push('1');
        }
        this.pushKeys(keys);
    }
    pushKeys(keys) {
        if (Array.isArray(keys)) {
            for (const key of keys) {
                this.#addKey(key, true);
            }
        }
        else {
            this.#addKey(keys, true);
        }
    }
}
exports.BasicCommandParser = BasicCommandParser;
